### DISCUSSION TITLE
How to get a strong handle from weak handle?

### DISCUSSION THREAD

**Discussion by Stranget0** [bevyengine/bevy#12735] [Q&A] [answered]
Hello!
So I have a HashMap which stores weak handles inside of it for each mesh enum which I use as key
The idea is that if the asset pointed by weak handle exist I can get the strong handle from it
If it doesn't, then I create, add the mesh and return it's strong handle
How can I achieve that? I don't see an obvious way to get strong handle from a weak one

**Comment by SkiFire13** [comment]
I believe you can use `AssetServer::get_id_handle` to this. Also, if you know all your handles will be weak ones you might as well directly store `AssetId`s.

**Reply by Stranget0** [reply]
Yeah, was thinking of this too reading from similar discussion, although I don't seem to be able to get a specific handle with specified asset id, everytime AssetServer is responding with None, not really sure what am I doing wrong here 😅
Here I tested this with logs and some relevant code:
```
2024-03-27T07:51:15.990252Z  ASSET SAVE key HexMapTile([0, 0, 0, 0, 0, 0]) id: AssetId<bevy_render::mesh::mesh::Mesh>{ index: 23198, generation: 0}
2024-03-27T07:51:15.990492Z   ASSETS MISS for key HexMapTile([0, 0, 0, 0, 0, 0]), id: AssetId<bevy_render::mesh::mesh::Mesh>{ index: 23198, generation: 0}
```

```
    fn get_or_create_mesh(
        &mut self,
        meshes: &mut ResMut<Assets<Mesh>>,
        mesh_type: &MeshType,
        layout: &HexLayout,
        asset_server: &Res<AssetServer>,
    ) -> Handle<Mesh> {
        let is_zero = mesh_type == &MeshType::HexMapTile([0, 0, 0, 0, 0, 0]);
        if let Some(handle) = self
            .meshes_map
            .get(mesh_type)
            .and_then(|asset_id| asset_server.get_id_handle(*asset_id))
        {
            if is_zero {
                info!("ASSETS HIT for key {:?} id: {}", mesh_type, handle.id());
            }
            return handle.clone();
        }
        if is_zero {
            if let Some(asset_id) = self.meshes_map.get(mesh_type) {
                info!("ASSETS MISS for key {:?}, id: {:?}", mesh_type, asset_id,);
            }
        }
        let mesh: Mesh = match mesh_type {
            // ...
        };

        let handle = meshes.add(mesh);
        self.meshes_map.insert(*mesh_type, handle.id());
        if is_zero {
            info!("ASSET SAVE key {:?} id: {}", mesh_type, handle.id());
        }
        handle
   }
```
        
I guess the asset server needs some time to process the data? I can get asset ids from my hashmap no problem, but asset server cannot find relevant handles

**Reply by Stranget0** [reply]
Update: 
I switched from `materials.add` to direct `asset_server.add` and now it is working. I guess updates in materials / meshes aren't tracked by asset_server and vice versa?

**Reply by SkiFire13** [reply]
Note that if you don't keep a strong handle somewhere your asset will be unloaded.

**Reply by Stranget0** [reply]
I have a strong handle in spawned entities and they don't lose their mesh, so I guess they should be fine :)
The returned `handle` is strong

**Reply by Stranget0** [reply]
Alright, I also made a test and store strong handles in a struct's vec after creation. The asset server still can't pick those up, when adding in meshes resource, but when adding directly to it by AssetServer::add it does 🤔

It seems that if I use AssetServer for adding all assets, including meshes, materials and images it would work, but is this really correct? Not sure if that could cause some problems as for example Res<Assets<Mesh>> don't seem to track what is happening in AssetServer and vice versa

**Reply by Stranget0** [reply, marked as answer]
So if anyone struggling with AssetServer not picking up immediately the addition to Assets I fixed this by using AssetServer::add instead of Assets::add and then doing AssetServer::get_id_handle
